File: c#/Days/Day1.cs
namespace aoc24.Days;

public static class Day1
{
    public class Input
    {
        internal List<int>[] Columns = [[], []];
    }

    public static Input ParseInput(string s)
    {
        var ret = new Input();
        using var sr = new StringReader(s);
        while (sr.ReadLine() is { } line)
        {
            foreach (var (n, i) in line.Split("   ").Select((n, i) => (n, i)))
            {
                ret.Columns[i].Add(int.Parse(n));
            }
        }
        return ret;
    }

    public static int Part1(Input input)
    {
        var list1 = input.Columns[0];
        var list2 = input.Columns[1];
        list1.Sort();
        list2.Sort();
        return list1.Zip(list2).Sum(pair => Math.Abs(pair.First - pair.Second));
    }

    public static int Part2(Input input)
    {
        var solution = 0;
        var cache = new Dictionary<int, int>();

        foreach (var i in input.Columns[0])
        {
            if (!cache.ContainsKey(i))
            {
                var count = input.Columns[1].Count(j => j == i);
                cache[i] = count;
            }

            solution += i * cache[i];
        }

        return solution;
    }
}

File: rust/src/days/day1.rs
use std::{convert::Infallible, str::FromStr};

#[derive(Debug)]
pub struct Input {
    columns: [Vec<i32>; 2],
}

impl FromStr for Input {
    type Err = Infallible;

    fn from_str(s: &str) -> Result<Self, Self::Err> {
        let mut columns: [Vec<i32>; 2] = std::array::from_fn(|_| Vec::new());
        s.lines().for_each(|line| {
            let (n1, n2) = line.split_once("   ").unwrap();
            columns[0].push(n1.parse().unwrap());
            columns[1].push(n2.parse().unwrap());
        });
        Ok(Input { columns })
    }
}

pub fn part1(input: Input) -> i32 {
    let mut v1 = input.columns[0].clone();
    let mut v2 = input.columns[1].clone();
    v1.sort();
    v2.sort();
    v1.into_iter().zip(v2).map(|(n1, n2)| (n1 - n2).abs()).sum()
}

pub fn part2(input: Input) -> i32 {
    let v1 = input.columns[0].clone();
    let v2 = input.columns[1].clone();
    v1.into_iter()
        .map(|n| n * v2.iter().filter(|&&i| i == n).count() as i32)
        .sum()
}

#[cfg(test)]
mod tests {
    use super::*;

    const INPUT: &str = "\
3   4
4   3
2   5
1   3
3   9
3   3";

    #[test]
    fn test_part1() {
        assert_eq!(11, part1(INPUT.parse().unwrap()));
    }

    #[test]
    fn test_part2() {
        assert_eq!(31, part2(INPUT.parse().unwrap()));
    }
}
